<!DOCTYPE html>
<html>
<head>
	<meta name="description" content="University final year project on single-cell RNA sequencing">
  <meta name="keywords" content="scRNA-seq, clustering, topological data analysis">
  <meta name="author" content="TomMakesThings">
	<meta name="viewport" content="width=device-width, initial-scale=1">

	<!-- Links to the CSS stylesheets -->
	<link rel="stylesheet" href="Stylesheets\main.css">
	<link rel="stylesheet" href="Stylesheets\navbar.css">
	<link rel="stylesheet" href="Stylesheets\about.css">
	<link rel="stylesheet" href="Stylesheets\datasets.css">
	<link rel="stylesheet" href="Stylesheets\clustering.css">
	<link rel="stylesheet" href="Stylesheets\tda.css">
	<link rel="stylesheet" href="Stylesheets\code.css">
	<link rel="stylesheet" href="Stylesheets\graph_finder.css">
	<link rel="stylesheet" href="Stylesheets\footer.css">
</head>

<body>

<div class="header">
	<div id="header-left" onmouseover="profileHover(this);" onmouseout="profileUnhover(this);">
		<a href="https://github.com/TomMakesThings"><img src="Assets/Images/Profile.jpg" alt="GitHub profile picture" id="profilepic"></a>
		<a href="https://tommakesthings.github.io/" id="username" class="navtext">TomMakesThings</a>
	</div>
	<div id="header-center">
		<a href="https://github.com/TomMakesThings/Clustering-and-TDA-of-scRNA-seq-Data" class="navtext">Clustering and TDA of scRNA-Seq Data</a>
	</div>
	<div class="header-right" id="visible-header">
		<a href="#about" class="navtext">About</a>
		<div class="nav-dropdown">
			<a href="#background" class="navtext">Background</a>
			<div class="nav-dropdown-content">
				<p><a href="#genes">Genes and Proteins</a></p>
				<p><a href="#scrnaseq">Single Cell RNA-Sequencing</a></p>
			</div>
		</div>
		<div class="nav-dropdown">
			<a href="#datasets" class="navtext">Data</a>
			<div class="nav-dropdown-content">
				<p><a href="#benchmark">Benchmark</a></p>
				<p><a href="#simulated">Simulated</a></p>
				<p><a href="#mousecortex">Mouse Cortex</a></p>
			</div>
		</div>
		<div class="nav-dropdown">
			<a href="#clustering" class="navtext">Clustering</a>
			<div class="nav-dropdown-content">
				<p><a href="#clusteringalgorithms">Clustering Algorithms</a></p>
				<p><a href="#preprocessing">Preprocessing</a></p>
				<p><a href="#measuringcorrectness">Measuring Correctness</a></p>
				<p><a href="#clusterresults">Results and Findings</a></p>
			</div>
		</div>
		<div class="nav-dropdown">
			<a href="#tda" class="navtext">TDA</a>
			<div class="nav-dropdown-content">
				<p><a href="#mapper">Mapper</a></p>
				<p><a href="#parametertuning">Parameter Tuning</a></p>
				<p><a href="#tdaresults">Results</a></p>
			</div>
		</div>
		<div class="nav-dropdown">
			<a href="#code" class="navtext">Code</a>
			<div class="nav-dropdown-content">
				<p><a href="#splatnotebook">Splat Simulator</a></p>
				<p><a href="#clusternotebook">Clustering and TDA</a></p>
			</div>
		</div>
		<a href="#graphfinder" class="navtext">Graphs</a>
	</div>
</div>

<div class="content">
	<div>
		<a name="about" class="anchor"></a>
  	<h1>About</h1>
		<p>This site gives an overview of my undergraduate dissertation, though may also be used as an introduction to some useful computational methods for the analysis of single-cell RNA sequencing data.
			The aim is to experiment with clustering and topological data analysis to detect hidden gene expression in three different types of datasets.
			For more detail, refer to the paper.
		</p>
	</div>

	<br>

	<div id="background-text">
		<a name="background" class="anchor"></a>
		<h1>Background</h1>
		<a name="genes" class="anchor"></a>
		<h2>Genes and Proteins</h2>
		<p>In biology, genes are small sections of DNA that encode proteins.
			These are important biomolecules with a large range of functions such as: enzymes, antibodies, structural proteins, storage or cell signalling.
			Studying gene expression can therefore reveal useful knowledge about the biological processes of cells.</p>
		<p>
			Proteins are produced in cells through two steps: transcription and translation.
			During transcription, DNA is unzipped and sections copied using the enzyme RNA polymerase to produce a complementary molecule called messenger RNA (mRNA).
			The mRNA then travels to a ribosome where a protein is constructed in a process known as translation.
			This means that cellular gene expression can be inferred through reading the sequence of bases RNA.</p>

		<a name="scrnaseq" class="anchor"></a>
		<h2>Single-Cell RNA-Sequencing</h2>
		<p><img src="Assets/Images/scRNA-seq-Steps.png" alt="Single-cell RNA sequencing steps" id="scRNA-seq-steps"></img></p>
		<p>Single-cell RNA sequencing (scRNA-seq) is a method in which genetic expression is measured across a population of heterogeneous cells of the same type.
			This has many uses such as detection of rare subpopulations of cells or gene co-expression.</p>
		<p>
			First viable, individual cells are isolated from a tissue sample.
			For each, the membrane is broken down in a process called lysis as this improves the capture of RNA molecules.
			Then tagged complementary DNA (cDNA) is synthesized from mRNA through reverse transcription.
			Small volumes of cDNA are amplified using a technique such as polymerase chain reaction (PCR) which facilitates rapid creation of millions of copies.
			Each cell’s cDNA library is pooled, then bases read and matched to genes to produce a dataset of counts for each cell.</p>
		<p>Improvements in sequencing technologies have made it easy for scientists to read gene expression for thousands to millions of cells.
			This provides an abundance of big data suitable for evaluation by computational methods that would be hard to analyse by hand in a lab.
		</p>
	</div>

	<br>

	<div>
		<a name="datasets" class="anchor"></a>
		<h1>Datasets</h1>
		<a name="benchmark" class="anchor"></a>
		<h2>Benchmark</h2>
		<p>The first dataset was created by Tian et al. and contains 16,468 gene counts for 902 cells produced using the 10X Chromium scRNA-seq protocol.
			It is officially named <a href="https://github.com/LuyiTian/sc_mixology/tree/master/data/csv" target="_blank">sc_10x</a>, though here is referred to as the "benchmark dataset" as it was used as a gold standard during testing and development.
		</p>

		<p><img src="Assets/Images/Benchmark-Counts.png" class="gene-counts" style="max-width: 770px"></p>

		<p>The cells consist of a mixture from three human lung adenocarcinoma cell lines HCC827, H1975 and H2228.
			These assignments are provided within the dataset's metadata and can therefore be used as target labels.
			However, it should be noted that for many real-world applications of scRNA-seq, these targets would be unknown.
			Each cell has been coloured accordingly in the plot below of PCA with two principal components.
			The 150 cells randomly selected for initial experimentation are represented by the coloured in dots.
		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Target_Groups/sc_10x_cell_lines_2D.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<a name="simulated" class="anchor"></a>
		<h2>Simulated</h2>
		<p>The second dataset consists of 2000 cells with 16,468 genes each.
			It was made during this project using the Splat simulator, which is part of the R package <a href="https://bioconductor.org/packages/devel/bioc/vignettes/splatter/inst/doc/splatter.html" target="_blank">Splatter</a>.
			The code and CSV files of the data are available to download from my <a href="https://github.com/TomMakesThings/Clustering-and-TDA-of-scRNA-seq-Data/tree/main" target="_blank">GitHub</a>.
			To replicate real gene expression, the Splat simulator was seeded with the benchmark dataset <a href="https://github.com/LuyiTian/sc_mixology/tree/master/data/csv" target="_blank">sc_10x</a>.
		</p>

		<p><img src="Assets/Images/Simulated-Counts.png" class="gene-counts" style="max-width: 830px;"></p>

		<p>Each sample belongs to one of four target groups as shown in the plot below.
			This time, 500 samples were randomly selected for initial experimentation.</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Target_Groups/splat_groups_2D.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<a name="mousecortex" class="anchor"></a>
		<h2>Mouse Cortex</h2>
		<p>The final dataset is by Zeisel et al. and contains 3005 brain cells each with 1072 genes from mouse cortex and hippocampus.
			It is available to download from <a href="http://linnarssonlab.org/cortex/" target="_blank">Linnarsson Lab</a>.
		</p>

		<p><img src="Assets/Images/Mouse-Cortex-Counts.png" class="gene-counts" style="max-width: 870px"></p>

		<p>Each cell is assigned to one of nine numerical groups and 47 subgroups determined using BackSPIN biclustering.
			 As explained in the <a href="https://science.sciencemag.org/content/347/6226/1138.abstract" target="_blank">original paper</a>, these subgroups comprise all known major cell types in the cortex.
			 The initial nine groups are labelled according to cell type as: astrocytes-ependymal, endothelial-mural, interneurons, microglia, oligodendrocytes, pyramidal CA1 or pyramidal SS.
			 Though groups 1 and 2 are assigned class joint pyramidal CA1, and 7 and 8 astrocytes-ependymal and so there are seven target labels.
			 Again 500 cells were randomly selected for initial experimentation.
		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Target_Groups/cortex_groups_2D.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

	</div>

	<br>

	<div>
		<a name="clustering" class="anchor"></a>
		<h1>Clustering</h1>

		<a name="clusteringalgorithms" class="anchor"></a>
		<h2>Clustering Algorithms</h2>
		<p>There are many clustering algorithms and tools to choose from, including those designed for scRNA-seq data such as SC3, SIMLR or BackSPIN.
			However, for this project I tested six standard clustering algorithms from <a href="https://scikit-learn.org/stable/modules/clustering.html" target="_blank">scikit-learn</a>:
		</p>
		<ul>
			<li><b>K-means</b>: A commonly used iterative centroid based algorithm that divides data points into k clusters</li>
			<li><b>Agglomerative hierarchical clustering</b>: A form of hierarchical clustering that builds a hierarchy of clusters by recursively merging pairs</li>
			<li><b>Balanced iterative reducing and clustering using hierarchies (BIRCH)</b>: Another form of hierarchical clustering designed for clustering large datasets</li>
			<li><b>Mini batch k-means</b>: An alternative version of k-means that instead clusters small subsets of the data each iteration to reduce memory usage</li>
			<li><b>Spectral clustering</b>: A type of clustering that constructs a connected graph of data points and assigns clusters by partitioning into subgraphs</li>
			<li><b>Gaussian mixture</b>: Clustering using a mixture of multiple Gaussian distributions</li>
		</ul>

		<a name="preprocessing" class="anchor"></a>
		<h2>Preprocessing</h2>
		<p>A problem with scRNA-seq data is that it is high-dimensional as each cell contains many features.
			Many of the genes only exhibit noise and so clustering will not work well if directly applied to the gene counts.
			Therefore the data must be preprocessed using a method to extract the most representational features.
		</p>
		<h3>Autoencoders</h3>

		<p id="autoencoder"><img src="Assets/Images/Autoencoder.png"></p>

		<p>An autoencoder is a type of artificial neural network with an encoder-decoder architecture that learns a compressed representation of the data.
			During training, the gene counts are given as input and the network will attempt to recreate this and return it at the output layer.
			After training, the output layer can be discarded and the encoding retrieved from the network's bottleneck layer.
			This learnt encoding can be used directly, or in combination with additional pre-processing techniques for clustering.
			It is good practise to test different network architectures and hyperparameters to improve the recreation accuracy of a dataset.
			For this project, the autoencoder's encoding size was set to 16 and the parameters optimised for the benchmark dataset.
			This includes:
		</p>
		<ul>
			<li>Setting the PReLU activation function after the first linear layer in the encoder</li>
  		<li>Using the AMSGrad variant of the Adam optimiser with learning rate 0.001</li>
  		<li>Training through mini-batch gradient descent with a batch size of four</li>
		</ul>

		<h3>Dimensionality Reduction Algorithms</h3>
		<p>There are lots of algorithms available for feature extraction and it is worth trying out different methods.
			These may be used in combination with, or as an alternative to the encoding produced by the autoencoder.
			Some commonly used techniques that I investigated are:
		</p>
		<ul>
			<li><b>Principal component analysis (PCA)</b>: A statistical technique that find the top principal components that maximise variance by solving a eigenvector problem</li>
			<li><b>Independent component analysis (ICA)</b>: A method designed for signal processing that transforms a set of vectors into a maximally independent set</li>
			<li><b>Non-negative matrix factorization (NMF)</b>: A technique that extracts features from a set of non-negative vectors</li>
		</ul>

	  <p class="feature-extractors"><img src="Assets/Images/Feature-Extractors.png"></p>

		<p>Before running these algorithms, the number of features to reduce the data to must be specified.
			Specific genes are not directly selected, but instead new features are created from weighted combinations of genes.
			Investigating which genes provide the largest contributions to these top components may also reveal additional information of interest.
			For example in the graph below, after applying PCA to the benchmark dataset, plotting the magnitude of the eigenvector loading scores for the top three principal components discloses which genes show the greatest variation between cells.
		</p>
		<p class="pca-scores"><img src="Assets/Images/Loading-Scores-Histogram.png"></p>

		<h3>Standardization</h3>
		<p>Standardization is the process of scaling features so that they have a mean of 0 and standard deviation 1.
			Applying this before feature extraction sometimes works very well at improving the performance of clustering as it can cause similar cells to be closer together while increasing distance between those that are dissimilar.
			It will also result in a more even distribution of gene contribution to extracted features, for example as shown in the graphs below.
			However, its effectiveness varies depending on the structure of data and the combination of other preprocessing methods applied and occasionally this step could be detrimental.
		</p>
		<p>Furthermore, it is recommended to apply standardization before running a clustering algorithm such as k-means, for example following feature extraction by PCA.
			This ensures all features are within the same range when the algorithm calculates distance.
		</p>

		<div class="heatmap-row">
			<div class="heatmap-column">
				<p><img src="Assets/Images/PCA-Loading-Scores.png"></p>
			</div>
			<div class="heatmap-column">
				<p><img src="Assets/Images/PCA-Loading-Scores-Standardized.png"></p>
			</div>
		</div>

		<h3>t-Distributed Stochastic Neighbor Embedding</h3>
		<p>t-SNE is an algorithm well suited to visualisation of high-dimensional data.
			It is popular for scRNA-seq data as it isolates groups of cells while preserving local similarities.
			This is achieved by minimising the Kullback-Leibler divergence over two probability distributions constructed from the high-dimensional and low dimensional space.
			Unlike other dimensionality reduction techniques, with t-SNE global geometry is lost due to the random initialisation of the algorithm.
			This means that the distance between and position of groups of cells do not have any significant meaning.
			If using t-SNE, it is advisable to apply another dimensionality reduction method beforehand to remove noise.
		</p>

		<p class="preprocessing"><img src="Assets/Images/Preprocessing.png"></p>
		<p class="preprocessing-cropped"><img src="Assets/Images/Preprocessing-Cropped.png"></p>

		<a name="measuringcorrectness" class="anchor"></a>
		<h2>Measuring Correctness</h2>
		<p>Clustering algorithms are unsupervised meaning that they are not guided by a target output.
			However, clustering can often be misleading and just because a cluster assignment looks reasonable does not mean it is correct.
		</p>

		<div class="cluster-row">
			<div class="cluster-column">
				<p><img src="Assets/Images/Clustering-Example.png" style="margin-left:-10px;"></p>
			</div>
			<div class="cluster-column">
				<p><img src="Assets/Images/Clustering-Example-2.png"></p>
			</div>
		</div>

		<p><img src="Assets/Images/Bipartite-Matching.png" id="bipartite"></p>

		<p>Arbitrary numerical values are outputted by clustering algorithms to represent the predicted cluster for each cell.
			As each selected dataset contains group labels, it possible to evaluate correctness by finding the optimal mapping between these labels and predictions.
			There are multiple methods to achieve this, but for this project I constructed a bipartite graph between the two sets and applied the Hungarian matching algorithm to find the maximal-weight matchings.
			These were used to calculate performance metrics such as accuracy and adjusted Rand index (ARI).
			The difference can also be plotted to visualise which cells the clustering algorithm was not able to correctly predict.
		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_E_68.7%25.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<a name="clusterresults" class="anchor"></a>
		<h2>Results and Findings</h2>

		<h3>Best Results</h3>
		<h4>Benchmark Dataset</h4>
		<p>One reason this dataset is ideal for benchmarking is that an extensive <a href="https://github.com/LuyiTian/sc_mixology/tree/master/data/benchmark_results/clustering" target="_blank">table of results</a> is avaliable from experimentation with different normalisation techiques, imputation methods and clustering algorithms commonly used for scRNA-seq including: RaceID, RaceID2, RCA, seurat, clusterExperiment, and SC3.
			These are also documented in a <a href="https://www.researchgate.net/publication/328071371_scRNA-seq_mixology_towards_better_benchmarking_of_single_cell_RNA-seq_protocols_and_analysis_methods" target="_blank">paper by L. Tian et al</a>.
			The best result they achieved was 0.742 ARI using SC3 clustering.
		</p>

		<p><img src="Assets/Images/Benchmark-ARI.png" id="benchmark-ari"></p>

		<p>By comparision, I was able to achieve a much higher ARI of 0.997 and accuracy 99.9% in which 901 out of 902 cells were assigned to the correct cell line.
			This was possible through a combination of standardization, PCA / ICA with three components and BIRCH / agglomerative hierarchical clustering.
		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_hierarchical_3_PCA_99.9%25.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<h4>Simulated Dataset</h4>
		<p>For the Splat simulated dataset, I was able to correctly group all 2000 cells resulting in 100% accuracy.
			This was accomplished through applying standardization and PCA / ICA with four components in combination with agglomerative hierarchical clustering, BIRCH or mini batch k-means.
		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Alternative_Algorithms/Simulated_Data/splat_birch_4_ICA_100%25.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<h4>Mouse Cortex Dataset</h4>
		<p>The best accuracy I was able to achieve for this dataset was poor at 44.4%.
			Compared to the previous two datasets, cells show an overlapping spectrum of gene expression in the lower dimensional space when using pre-processing techniques attempted in this project.
			This is a problem as standard clustering methods require some degree of seperation to reliably assign clusters.
		</p>
		<p>Another issue is that the ground truth was derived using the BackSPIN biclustering algorithm.
			While standard clustering algorithms such as k-means divide cells into clusters, biclustering is a different class of algorithms that cluster cells and genes simultaneously enabling cells to be grouped by gene co-expression.
			As explained in the <a href="https://science.sciencemag.org/content/347/6226/1138.abstract" target="_blank">original paper by Zeisel et al.</a>, many of the genes in this dataset are not related to every sample making biclustering more suitable.
		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Hierarchical_Clustering/Real_Evaluation_Data/cortex_hierarchical_9_PCA_SU_44%25.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<h3>Observations</h3>
		<h4>Autoencoders</h4>
		<ul>
			<li>The best clustering results for the benchmark dataset were attained without encoding, while top results were achieved both with and without encoding the simulated data.
			Effectiveness was therefore dependent on the data, though it is always worth testing multiple approaches.</li>
			<li>I kept the encoding size at 16 for this project, though ideally different encoding sizes should be tested.</li>
			<li>t-SNE was more effective when combined with the autoencoder's encoding.</li>
		</ul>

		</p>

		<h4>t-SNE</h4>
		<p>This is a good option to try if cells are not separable by other methods, however, should be used with caution as it has can produce misleading results.
			For example in the graph below, all cells in the predicted clusters are close together so without looking at the expected cell lines it is not obvious that one cell is incorrect.
		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_SE_t-SNE_99.3%25.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<p>If using t-SNE, different values for the hyperparameter perplexity should be tested as changing this causes results to vary significantly.
		</p>

		<p><img src="Assets/Images/Perplexity.png" id="perplexity"></p>

	</div>

	<br>

	<div>
		<a name="tda" class="anchor"></a>
		<h1>Topological Data Analysis</h1>
		<a name="mapper" class="anchor"></a>
		<h2>Mapper</h2>
		<p>A simplicial complex is a type of graph that represents data with many features through a high-dimensional shape.
			Compare to graphs produced through standard dimensionality reduction techniques, this gives a higher-dimensional overview of a dataset to provide insight into relationships in gene expression between cells.
		</p>
		<p>Mapper is a TDA algorithm that constructs a simplicial complex through dimensionality reduction, clustering, and graph networks.
			<a href="https://kepler-mapper.scikit-tda.org/en/latest/" target="_blank">Kepler Mapper</a> is a Python implementation of Mapper that was used for this project.
		</p>

		<a name="parametertuning" class="anchor"></a>
		<h2>Parameter Tuning</h2>
		<p>There is no right way to construct a simplicial complex, but the aim is to reveal topological features.
			For Kepler Mapper, the main hyperparameters to tune are epsilon and hypercubes:
		</p>
		<ul>
			<li><b>Epsilon</b>: When epsilon is small, many of the nodes are separated, while increasing it too much results in most nodes being connected losing the topological features.</li>
			<li><b>Hypercubes</b>: Using only a few hypercubes creates a simple graph, while increasing this reveals more complexity.</li>
		</ul>

		<p class="tda-graph"><img src="Assets/Images/TDA-Epsilon.png"></p>

		<p class="tda-graph"><img src="Assets/Images/TDA-Hypercubes.png"></p>

		<a name="tdaresults" class="anchor"></a>
		<h2>Results</h2>
		<p>Each dataset was first projected onto a lower-dimensional space using 2 principal components, then a simplicial complex constructed using 10 hypercubes with 0.15 overlap and k-means clustering.
			For all three datasets, the same hyperparameters were used so that topological features can be compared.</p>
		<h3>Simplicial Complex of Benchmark Dataset</h3>
		<p>In the interactive graph below, the cell lines are identifiable as the three most connected regions.
			As most nodes are connected, this suggests that most cells have some correlation in gene expression.
			There are also several nodes that are not joined to the main complex representing small subgroups of cells with more unusual gene expression.
			However, it is worth pointing out that if the parameter epsilon was increased then some of these nodes would join the main complex, while if it was decreased then more of these distinct nodes would form.
		</p>
		<div class="mapper-graph" style="height: 1600px; margin-bottom: -950px;">
			<iframe src="Assets/IFrame_HTML/benchmarking_scRNAseq_graph.html" class="mapper-graph-html" scrolling="no"></iframe>
		</div>

		<h3>Simplicial Complex of Simulated Dataset</h3>
		<p>Compared to the benchmarking data, the groups of the simulated data are much more clearly defined and separated from one another.
			This suggests that gene expression between the groups do not overlap as much as true biological data.
			Furthermore, the groups themselves form straight lines indicating that there is a lot less variation within each of them.
		</p>
		<div class="mapper-graph" style="height: 1450px; margin-bottom: -860px;">
			<iframe src="Assets/IFrame_HTML/simulated_scRNAseq_graph.html" class="mapper-graph-html" scrolling="no"></iframe>
		</div>

		<h3>Simplicial Complex of Mouse Cortex Dataset</h3>
		<p>The simplicial complex of this dataset is far more complicated than the others reflecting the greater variation of gene expression within the mouse cortex and hippocampus.
			As seen during clustering, the groups are not well defined, though this graph implies most cells can be split into two distinct regions.
			There are also far more individual nodes compared to the other datasets, though this is likely because this dataset contains the most cells.
		</p>
		<div class="mapper-graph" style="height: 2250px; margin-bottom: -1360px;">
			<iframe src="Assets/IFrame_HTML/evaluation_scRNAseq_graph.html" class="mapper-graph-html" scrolling="no"></iframe>
		</div>

	</div>

	<br>

	<div>
		<a name="code" class="anchor"></a>
		<h1>Code</h1>
		<p>Code was written in Python 3.6.9 and R 4.0.3 across two Jupyter notebooks, both of which are available to download from <a href="https://github.com/TomMakesThings/Clustering-and-TDA-of-scRNA-seq-Data/tree/main/Jupyter_Notebooks" target="_blank" rel="noopener noreferrer">GitHub</a>.
			These were developed in <a href="https://colab.research.google.com/notebooks/intro.ipynb?utm_source=scs-index#recent=true" target="_blank">Google Colab</a>, though alternatively can be run through <a href="https://jupyter.org/install.html" target="_blank">Jupyter Notebook</a>.
		</p>

		<a name="splatnotebook" class="anchor"></a>
		<h2>Splat Simulator</h2>
		<p>The purpose of notebook <a href="https://github.com/TomMakesThings/Clustering-and-TDA-of-scRNA-seq-Data/blob/main/Jupyter_Notebooks/Splat_Simulator.ipynb" target="_blank">Splat_Simulator.ipynb</a> is to produce new, artificial scRNA-seq data.
			For this project, it was used to create the <a href="#simulated">simulated dataset</a>, though can easily be altered to make new data for other purposes.
		</p>
		<p>
			Gene counts and group labels are generated using the Splat simulator, which is part of the R package <a href="https://bioconductor.org/packages/devel/bioc/vignettes/splatter/inst/doc/splatter.html" target="_blank">Splatter</a>, and so the code contains a mix between Python and inline R.
			To mimick true biological gene expression, the benchmark dataset has been set to use as a seed, though this could be swapped out to imitate another dataset.
			After seeding the simulator, datapoints are generated with each belonging to one of four groups.
			The new data and labels are then saved as CSV files using Python.
			These can be downloaded and reopened to use in the second notebook.
		</p>
		<div class="notebook">
			<iframe src="Assets/IFrame_HTML/Splat_Simulator.html" class="notebook-html"></iframe>
		</div>

		<a name="clusternotebook" class="anchor"></a>
		<h2>Clustering and TDA</h2>
		<p>In the notebook <a href="https://github.com/TomMakesThings/Clustering-and-TDA-of-scRNA-seq-Data/blob/main/Jupyter_Notebooks/Clustering_and_TDA.ipynb" target="_blank">Clustering_and_TDA.ipynb</a>, experimentation is performed on the three given datasets.
			However, it has been designed so that it could be easily run with new data.
		</p>
		<p>
			First the datasets and their target labels are opened as dataframes.
			These datasets are downloaded from URL so that the notebook can be run with no set up required.
			Next, a dataset is selected and an autoencoder with customisable hyperparameters created using PyTorch Lightning to use as a feature extractor for the gene counts.
			Then clustering is performed to divide cells into groups which show similar gene expression.
			Several clustering algorithms can be chosen including: k-means, agglomerative hierarchical, BIRCH, mini-batch k-means, spectral and Gaussian mixture.
			The encoding produced by the autoencoder can optionally be used, along with other dimensionality reduction methods such as PCA, ICA or NMF and techniques such as standardization and t-SNE.
			At the end of the notebook, Kepler Mapper is run on the gene counts to produce a simplicial complex to reveal the topological shape of the high-dimensional data.
		</p>
		<div class="notebook">
			<iframe src="Assets/IFrame_HTML/Clustering_and_TDA.html" class="notebook-html"></iframe>
		</div>

	</div>

	<br>

	<div id="graph-finder">
		<a name="graphfinder" class="anchor"></a>
		<h1>Graph Finder</h1>
		<p>Enter keywords to search for a clustering or topological data analysis experiment result:</p>

		<div id="graphsearch">
			<input type="search" name="search" placeholder="Search..." id="query" autocomplete="off" onclick="showDropdown()" oninput="showDropdown()" onkeyup="searchFilter()" onsearch="searchFilter()" aria-label="Find a graph">
			<div id="dropdown-content">
				<a class="dropdown-category" id="cluster-category">Clustering</a>
				<!--Benchmark clustering graphs-->
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_E_68.7%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, autoencoder encoding, 68.7% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_SE_97.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this) ; hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, standardization, autoencoder encoding, 97.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_SE_t-SNE_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this) ; hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, standardization, autoencoder encoding, t-SNE with perplexity 30, 99.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_SU_72.7%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, standardization, no encoding, 72.7% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_U_64%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, no encoding, 64% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_3_PCA_E_98.7%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 3 principal components, autoencoder encoding, 98.7% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_3_PCA_SE_98%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 3 principal components, standardization, autoencoder encoding, 98% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_3_PCA_SU_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 3 principal components, standardization, no encoding, 99.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_3_PCA_U_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 3 principal components, no encoding, 100% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_4_PCA_E_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 4 principal components, autoencoder encoding, 99.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_7_PCA_SE_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 7 principal components, standardization, autoencoder encoding, 99.3% accuracy</a>

				<a class="cluster-link link-option" data-href="Graphs/Hierarchical_Clustering/Benchmarking_Data/sc_10x_hierarchical_2_PCA_SE_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>Agglomerative Hierarchical Clustering on Benchmark Data:</b> 150 cells, PCA with 2 principal components, standardization, autoencoder encoding, 99.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Hierarchical_Clustering/Benchmarking_Data/sc_10x_hierarchical_3_PCA_SU_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian cluster component analysis">
				<b>Agglomerative Hierarchical Clustering on Benchmark Data:</b> 150 cells, PCA with 3 principal components, standardization, no encoding, 100% accuracy</a>

				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_hierarchical_3_PCA_99.9%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian cluster component analysis">
				<b>Agglomerative Hierarchical Clustering on Benchmark Data:</b> 902 cells, PCA with 3 principal components, standardization, no encoding, 99.9% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_hierarchical_3_ICA_99.9%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian cluster independent component analysis independant">
				<b>Agglomerative Hierarchical Clustering on Benchmark Data:</b> 902 cells, ICA with 4 independent components, standardization, no encoding, 99.9% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_k-means_3_NMF_87%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster non negative matrix factorization">
				<b>K-Means on Benchmark Data:</b> 902 cells, NMF with 3 basis components, standardization, no encoding, 87% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_gaussian_4_NMF_99.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian cluster non negative matrix factorization">
				<b>Gaussian Mixture on Benchmark Data:</b> 902 cells, NMF with 4 basis components, standardization, no encoding, 99.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_minibatch_k-means_2_PCA_t-SNE_99%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>Minibatch K-Means on Benchmark Data:</b> 902 cells, PCA with 2 principal components, standardization, autoencoder encoding, t-SNE with 30 perplexity, 99% accuracy</a>

				<!--Simulated clustering graphs-->
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_2_PCA_E_49.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 2 principal components, autoencoder encoding, 49.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_2_PCA_SE_92.4%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 2 principal components, standardization, autoencoder encoding, 92.4% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_2_PCA_SU_53.4%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 2 principal components, standardization, no encoding, 53.4% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_2_PCA_U_48.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 2 principal components, no encoding, 48.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_6_PCA_E_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 6 principal components, autoencoder encoding, 100% accuracy</a>

				<a class="cluster-link link-option" data-href="Graphs/Hierarchical_Clustering/Simulated_Data/splat_hierarchical_6_PCA_E_99.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom cluster analysis">
				<b>Agglomerative Hierarchical Clustering on Simulated Data:</b> 500 cells, PCA with 6 principal components, autoencoder encoding, 99.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_birch_4_PCA_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis balanced iterative reducing and clustering using hierarchies">
				<b>BIRCH on Simulated Data:</b> 2000 cells, PCA with 4 principal components, standardization, no encoding, 100% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_birch_4_ICA_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis balanced iterative reducing and clustering using hierarchies">
				<b>BIRCH on Simulated Data:</b> 2000 cells, ICA with 4 independent components, standardization, no encoding, 100% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_hierarchical_11_NMF_85%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis cluster non negative matrix factorization">
				<b>Agglomerative Hierarchical on Simulated Data:</b> 2000 cells, NMF with 11 basis components, no encoding, 85% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_spectral_2_PCA_t-SNE_50%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis cluster tsne stochastic neighbor embedding">
				<b>Spectral Clustering on Simulated Data:</b> 2000 cells, PCA with 2 principal components, autoencoder encoding, t-SNE with perplexity 30, 50% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_spectral_2_PCA_t-SNE_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis cluster tsne stochastic neighbor embedding">
				<b>Spectral Clustering on Simulated Data:</b> 2000 cells, PCA with 2 principal components, autoencoder encoding, t-SNE with perplexity 30, 100% accuracy</a>

				<!--Mouse cortex clustering graphs-->
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_2_PCA_SU_36.6%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 2 principal components, standardization, no encoding, 36.6% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_2_PCA_U_29.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 2 principal components, no encoding, 29.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_15_PCA_SE_36.4%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 15 principal components, standardization, autoencoder encoding, 36.4% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_16_PCA_SU_t-SNE_33.2%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 30 perplexity, 33.2% accuracy, </a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=5.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 5 perplexity, 30% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=10.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 10 perplexity, 28.4% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=20.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 20 perplexity, 29.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=100.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 100 perplexity, 34% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=300.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 300 perplexity, 37% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=600.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 600 perplexity, 21.6% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Hierarchical_Clustering/Real_Evaluation_Data/cortex_hierarchical_9_PCA_SU_44%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis">
				<b>Agglomerative Hierarchical Clustering on Mouse Cortex Data:</b> 500 cells, PCA with 9 principal components, standardization, no encoding, 44% accuracy</a>

				<!--TDA graphs-->
				<a class="dropdown-category" id="tda-category">Topological Data Analysis</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Parameter_Optimisation/scRNAseq_graph_ϵ=0.05.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian simplicial complex kepler mapper tda topological data analysis">
				<b>Benchmark Data</b>: Mapper simplicial complex with epsilon 0.05 and 10 cubes</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Parameter_Optimisation/scRNAseq_graph_ϵ=0.2.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian simplicial complex kepler mapper tda topological data analysis">
				<b>Benchmark Data</b>: Mapper simplicial complex with epsilon 0.2 and 10 cubes</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Parameter_Optimisation/scRNAseq_graph_ϵ=0.4.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian simplicial complex kepler mapper tda topological data analysis">
				<b>Benchmark Data</b>: Mapper simplicial complex with epsilon 0.4 and 10 cubes</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Parameter_Optimisation/scRNAseq_graph_ϵ=0.8.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian simplicial complex kepler mapper tda topological data analysis">
				<b>Benchmark Data</b>: Mapper simplicial complex with epsilon 0.8 and 10 cubes</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Final_Graphs/benchmarking_scRNAseq_graph.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian simplicial complex kepler mapper tda topological data analysis">
				<b>Benchmark Data</b>: Mapper simplicial complex with epsilon 0.15 and 10 cubes</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Final_Graphs/simulated_scRNAseq_graph.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splatter tom fake generated simplicial complex kepler mapper tda topological data analysis">
				<b>Simulated Data</b>: Mapper simplicial complex with epsilon 0.15 and 10 cubes</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Final_Graphs/evaluation_scRNAseq_graph.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna simplicial complex kepler mapper tda topological data analysis">
				<b>Mouse Cortex Data</b>: Mapper simplicial complex with epsilon 0.15 and 10 cubes</a>
			</div>
		</div>

		<br>
		<br>
		<br>

		<p id="graph-selected"></p>

		<div id="graph-wrapper">
			<iframe src="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_E_68.7%25.html" id="graph-html" scrolling="no"></iframe>
		</div>

	</div>

</div>

<br>

<div class="footer">
	<div class="footer-center">
		<a href="https://github.com/TomMakesThings">&copy; TomMakesThings <script>document.write(new Date().getUTCFullYear());</script></a>
	</div>
</div>

<script src="Javascript/graph_search.js"></script>
<script src="Javascript/navbar.js"></script>

</body>
</html>
