<!DOCTYPE html>
<html>
<head>
	<meta name="description" content="University final year project on single-cell RNA sequencing">
  <meta name="keywords" content="scRNA-seq, clustering, topological data analysis">
  <meta name="author" content="TomMakesThings">
	<meta name="viewport" content="width=device-width, initial-scale=1">

	<!-- Links to the CSS stylesheets -->
	<link rel="stylesheet" href="Stylesheets\main.css">
	<link rel="stylesheet" href="Stylesheets\navbar.css">
	<link rel="stylesheet" href="Stylesheets\about.css">
	<link rel="stylesheet" href="Stylesheets\datasets.css">
	<link rel="stylesheet" href="Stylesheets\graph_finder.css">
	<link rel="stylesheet" href="Stylesheets\footer.css">
</head>

<body>

<div class="header">
	<div id="header-left">
		<a href="https://github.com/TomMakesThings"><img src="Assets/Images/Profile.jpg" alt="GitHub profile picture" id="profilepic" onmouseover="profileHover(this);" onmouseout="profileUnhover(this);"></a>
		<a href="https://tommakesthings.github.io/" id="username" class="navtext">TomMakesThings</a>
	</div>
	<div id="header-center">
		<a href="index.html" class="navtext">Clustering and TDA of scRNA-Seq Data</a>
	</div>
	<div class="header-right">
		<a href="#about" class="navtext">About</a>
		<a href="#background" class="navtext">Background</a>
		<a href="#datasets" class="navtext">Data</a>
		<a href="#clustering" class="navtext">Clustering</a>
		<a href="#code" class="navtext">Code</a>
		<a href="#graphfinder" class="navtext">Graphs</a>
	</div>
</div>

<div class="content">
	<div>
		<a name="about" class="anchor"></a>
  	<h1>About</h1>
		<p>This site gives a brief overview of my undergraduate dissertation, though may also be used as an introduction to some useful computational methods for the analysis of single-cell RNA sequencing data.
			The aim is to experiment with clustering and topological data analysis to detect hidden gene expression in three different types of datasets.
			For more information, read the paper.
		</p>
	</div>

	<br>

	<div id="background-text">
		<a name="background" class="anchor"></a>
		<h1>Background</h1>
		<h2>Genes and Proteins</h2>
		<p>In biology, genes are small sections of DNA that encode proteins.
			These are important biomolecules with a large range of functions such as: enzymes, antibodies, structural proteins, storage or cell signalling.
			Studying gene expression can therefore reveal useful knowledge about the biological processes of cells.</p>
		<p>
			Proteins are produced in cells through two steps: transcription and translation.
			During transcription, DNA is unzipped and sections copied using the enzyme RNA polymerase to produce a complementary molecule called messenger RNA (mRNA).
			The mRNA then travels to a ribosome where a protein is constructed in a process known as translation.
			This means that cellular gene expression can be inferred through reading the sequence of bases RNA.</p>
		<h2>Single-Cell RNA-Sequencing</h2>
		<p><img src="Assets/Images/scRNA-seq-Steps.png" alt="Single-cell RNA sequencing steps" id="scRNA-seq-steps"></img></p>
		<p>Single-cell RNA sequencing (scRNA-seq) is a method in which genetic expression is measured across a population of heterogeneous cells of the same type.
			This has many uses such as detection of rare subpopulations of cells or gene co-expression.</p>
		<p>
			First viable, individual cells are isolated from a tissue sample.
			For each, the membrane is broken down in a process called lysis as this improves the capture of RNA molecules.
			Then tagged complementary DNA (cDNA) is synthesized from mRNA through reverse transcription.
			Small volumes of cDNA are amplified using a technique such as polymerase chain reaction (PCR) which facilitates rapid creation of millions of copies.
			Each cellâ€™s cDNA library is pooled, then bases read and matched to genes to produce a dataset of counts for each cell.</p>
		<p>Improvements in sequencing technologies have made it easy for scientists to read gene expression for thousands to millions of cells.
			This provides an abundance of big data suitable for evaluation by computational methods that would be hard to analyse by hand in a lab.
		</p>
	</div>

	<br>

	<div>
		<a name="datasets" class="anchor"></a>
		<h1>Datasets</h1>
		<h2>Benchmark</h2>
		<p>The first dataset was created by Tian et al. and contains 16,468 gene counts for 902 cells produced using the 10X Chromium scRNA-seq protocol.
			It is officially named <a href="https://github.com/LuyiTian/sc_mixology/tree/master/data/csv" target="_blank">sc_10x</a>, though here is referred to as the "benchmark dataset" as it was used as a gold standard during testing and development.
		</p>

		<p><img src="Assets/Images/Benchmark-Counts.png" class="gene-counts" style="max-width: 900px"></p>

		<p>The cells consist of a mixture from three human lung adenocarcinoma cell lines HCC827, H1975 and H2228.
			These assignments are provided within the dataset's metadata and can therefore be used as target labels.
			However, it should be noted that for many real-world applications of scRNA-seq, these targets would be unknown.
			Each cell has been coloured accordingly in the plot below of PCA with two principal components.
			The 150 cells randomly selected for initial experimentation can be seen by clicking on second legend on the graph.
		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Target_Groups/sc_10x_cell_lines_2D.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<h2>Simulated</h2>
		<p>The second dataset was made during this project and is available to download from my <a href="https://github.com/TomMakesThings/Clustering-and-TDA-of-scRNA-seq-Data/tree/main/Data/Datasets">GitHub</a>.
			It was created using the Splat simulator, which is part of the R package <a href="https://bioconductor.org/packages/devel/bioc/vignettes/splatter/inst/doc/splatter.html" target="_blank">Splatter</a>,
			and consists 2000 cells with 16,468 genes each.
			To replicate real gene expression, the Splat simulator was seeded with the benchmark dataset sc_10x.
		</p>

		<p><img src="Assets/Images/Simulated-Counts.png" class="gene-counts" style="max-width: 950px"></p>

		<p>Each sample belongs to one of four target groups as shown below.
			This time, 500 samples were initially selected.</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Target_Groups/splat_groups_2D.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

		<h2>Mouse Cortex</h2>
		<p>The final dataset is by Zeisel et al. and contains 3005 brain cells each with 1072 genes from mouse cortex and hippocampus.
			It is available to download from <a href="http://linnarssonlab.org/cortex/">Linnarsson Lab</a>.
		</p>

		<p><img src="Assets/Images/Mouse-Cortex-Counts.png" class="gene-counts" style="max-width: 900px"></p>

		<p>Each cell is assigned to one of nine groups and 47 subgroups determined using BackSPIN biclustering.
			 As explained in the <a href="https://science.sciencemag.org/content/347/6226/1138.abstract">original paper</a>, these subgroups comprise all known major cell types in the cortex.

		</p>

		<div class="dataset-graph">
			<iframe src="Graphs/Target_Groups/cortex_groups_2D.html" class="dataset-graph-html" scrolling="no"></iframe>
		</div>

	</div>

	<br>

	<div>
		<a name="clustering" class="anchor"></a>
		<h1>Clustering</h1>
		<p>The goal of clustering is to group cells that show similar gene expression.

			However a problem with scRNA-seq data is that it is high-dimensional as each cell contains many features.
			This means that clustering will not work well if directly applied.
			Therefore, a method to extract and / or combine the most representational features is required.
		</p>

		<h2>Autoencoder</h2>
		<p>T</p>


	</div>

	<br>

	<div>
		<a name="code" class="anchor"></a>
		<h1>Code</h1>
		<p>This
		</p>
		<a href="Notebooks/Splat_Simulator.html" target="_blank" rel="noopener noreferrer">Notebooks</a>
		<a href="Notebooks/Clustering_and_TDA.html" target="_blank" rel="noopener noreferrer">Notebooks</a>
		<a href="https://github.com/TomMakesThings/Clustering-and-TDA-of-scRNA-seq-Data" target="_blank" rel="noopener noreferrer">Repository</a>
	</div>

	<br>

	<div id="graph-finder">
		<a name="graphfinder" class="anchor"></a>
		<h1>Graph Finder</h1>
		<p>Enter keywords to search for a clustering or topological data analysis experiment result:</p>

		<form id="graphform" role="search" autocomplete="off">
			<input type="search" name="search" placeholder="Search..." id="query" onclick="showDropdown()" oninput="showDropdown()" onkeyup="searchFilter()" onsearch="searchFilter()" aria-label="Find a graph">
			<div id="dropdown-content">
				<a class="dropdown-category" id="cluster-category">Clustering</a>
				<!--Benchmark clustering graphs-->
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_E_68.7%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, autoencoder encoding, 68.7% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_SE_97.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this) ; hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, standardization, autoencoder encoding, 97.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_SU_72.7%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, standardization, 72.7% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_U_64%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 2 principal components, no encoding, 64% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_3_PCA_E_98.7%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 3 principal components, autoencoder encoding, 98.7% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_3_PCA_SE_98%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 3 principal components, standardization, autoencoder encoding, 98% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_3_PCA_SU_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 3 principal components, standardization, no encoding, 99.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_3_PCA_U_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 3 principal components, no encoding, 100% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_4_PCA_E_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 4 principal components, autoencoder encoding, 99.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_7_PCA_SE_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>K-Means on Benchmark Data:</b> 150 cells, PCA with 7 principal components, standardization, autoencoder encoding, 99.3% accuracy</a>

				<a class="cluster-link link-option" data-href="Graphs/Hierarchical_Clustering/Benchmarking_Data/sc_10x_hierarchical_2_PCA_SE_99.3%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis">
				<b>Agglomerative Hierarchical Clustering on Benchmark Data:</b> 150 cells, PCA with 2 principal components, standardization, autoencoder encoding, 99.3% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Hierarchical_Clustering/Benchmarking_Data/sc_10x_hierarchical_3_PCA_SU_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian cluster component analysis">
				<b>Agglomerative Hierarchical Clustering on Benchmark Data:</b> 150 cells, PCA with 3 principal components, standardization, no encoding, 100% accuracy</a>

				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_hierarchical_3_PCA_99.9%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian cluster component analysis">
				<b>Agglomerative Hierarchical Clustering on Benchmark Data:</b> 902 cells, PCA with 3 principal components, standardization, no encoding, 99.9% accuracy, 0.997 ARI</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_hierarchical_3_ICA_99.9%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian cluster independent component analysis independant">
				<b>Agglomerative Hierarchical Clustering on Benchmark Data:</b> 902 cells, ICA with 4 independent components, standardization, no encoding, 99.9% accuracy, 0.997 ARI</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_k-means_3_NMF_87%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster non negative matrix factorization">
				<b>K-Means on Benchmark Data:</b> 902 cells, NMF with 3 basis components, standardization, no encoding, 87% accuracy, 0.696 ARI</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_gaussian_4_NMF_99.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian cluster non negative matrix factorization">
				<b>Gaussian Mixture on Benchmark Data:</b> 902 cells, NMF with 4 basis components, standardization, no encoding, 99.8% accuracy, 0.993 ARI</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Benchmarking_Data/sc_10x_minibatch_k-means_2_PCA_t-SNE_99%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian kmeans cluster component analysis tsne">
				<b>Minibatch K-Means on Benchmark Data:</b> 902 cells, PCA with 2 principal components, standardization, autoencoder encoding, t-SNE with 30 perplexity, 99% accuracy, 0.970 ARI</a>

				<!--Simulated clustering graphs-->
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_2_PCA_E_49.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 2 principal components, autoencoder encoding, 49.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_2_PCA_SE_92.4%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 2 principal components, standardization, autoencoder encoding, 92.4% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_2_PCA_SU_53.4%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 2 principal components, standardization, no encoding, 53.4% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_2_PCA_U_48.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 2 principal components, no encoding, 48.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Simulated_Data/splat_k-means_6_PCA_E_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom kmeans cluster component analysis">
				<b>K-Means on Simulated Data:</b> 500 cells, PCA with 6 principal components, autoencoder encoding, 100% accuracy</a>

				<a class="cluster-link link-option" data-href="Graphs/Hierarchical_Clustering/Simulated_Data/splat_hierarchical_6_PCA_E_99.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom cluster analysis">
				<b>Agglomerative Hierarchical Clustering on Simulated Data:</b> 500 cells, PCA with 6 principal components, autoencoder encoding, 99.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_birch_4_PCA_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis balanced iterative reducing and clustering using hierarchies">
				<b>BIRCH on Simulated Data:</b> 2000 cells, PCA with 4 principal components, standardization, no encoding, 100% accuracy, ARI 1.0</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_birch_4_ICA_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis balanced iterative reducing and clustering using hierarchies">
				<b>BIRCH on Simulated Data:</b> 2000 cells, ICA with 4 independent components, standardization, no encoding, 100% accuracy, ARI 1.0</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_hierarchical_11_NMF_85%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis cluster non negative matrix factorization">
				<b>Agglomerative Hierarchical on Simulated Data:</b> 2000 cells, NMF with 11 basis components, no encoding, 85% accuracy, ARI 0.857</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_spectral_2_PCA_t-SNE_50%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis cluster tsne">
				<b>Spectral Clustering on Simulated Data:</b> 2000 cells, PCA with 2 principal components, autoencoder encoding, 50% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/Alternative_Algorithms/Simulated_Data/splat_spectral_2_PCA_t-SNE_100%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splat fake generated tom analysis cluster tsne">
				<b>Spectral Clustering on Simulated Data:</b> 2000 cells, PCA with 2 principal components, autoencoder encoding, 100% accuracy</a>

				<!--Mouse cortex clustering graphs-->
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_2_PCA_SU_36.6%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 2 principal components, standardization, no encoding, 36.6% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_2_PCA_U_29.8%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 2 principal components, no encoding, 29.8% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_15_PCA_SE_36.4%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 15 principal components, standardization, autoencoder encoding, 36.4% accuracy</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_16_PCA_SU_t-SNE_33.2%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 30 perplexity, 33.2% accuracy, </a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=5.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 5 perplexity, 30% accuracy, 0.104 ARI, 0.451 silhouette coefficient</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=10.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 10 perplexity, 28.4% accuracy, 0.1 ARI, 0.435 silhouette coefficient</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=20.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 20 perplexity, 29.8% accuracy, 0.112 ARI, 0.436 silhouette coefficient</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=100.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 100 perplexity, 34% accuracy, 0.125 ARI, 0.374 silhouette coefficient</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=300.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 300 perplexity, 37% accuracy, 0.137 ARI, 0.399 silhouette coefficient</a>
				<a class="cluster-link link-option" data-href="Graphs/K-Means/Real_Evaluation_Data/cortex_k-means_t-SNE_perplexity=600.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis tsne stochastic neighbor embedding">
				<b>K-Means on Mouse Cortex Data:</b> 500 cells, PCA with 16 principal components, standardization, no encoding, t-SNE with 600 perplexity, 21.6% accuracy, 0.005 ARI, 0.528 silhouette coefficient</a>
				<a class="cluster-link link-option" data-href="Graphs/Hierarchical_Clustering/Real_Evaluation_Data/cortex_hierarchical_9_PCA_SU_44%25.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna kmeans cluster component analysis">
				<b>Agglomerative Hierarchical Clustering on Mouse Cortex Data:</b> 500 cells, PCA with 9 principal components, standardization, no encoding, 44% accuracy</a>

				<!--TDA graphs-->
				<a class="dropdown-category" id="tda-category">Topological Data Analysis</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Final_Graphs/benchmarking_scRNAseq_graph.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="sc_10x chromium lung adenocarcinoma luyi tian simplicial complex kepler mapper tda topological data analysis">
				Benchmark</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Final_Graphs/simulated_scRNAseq_graph.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="splatter tom fake generated simplicial complex kepler mapper tda topological data analysis">Simulated</a>
				<a class="tda-link link-option" data-href="Graphs/Topological_Data_Analysis/Final_Graphs/evaluation_scRNAseq_graph.html"
				target="_blank" rel="noopener noreferrer" onclick="displayGraph(this); hideDropdown()" data-tags="zeisel hippocampus mrna simplicial complex kepler mapper tda topological data analysis">Mouse Cortex</a>
			</div>
		</form>

		<br>
		<br>
		<br>

		<a id="graph-selected"></a>

		<div id="graph-wrapper">
			<iframe src="Graphs/K-Means/Benchmarking_Data/sc_10x_k-means_2_PCA_E_68.7%25.html" id="graph-html" scrolling="no"></iframe>
		</div>

	</div>

</div>

<br>

<div class="footer">
	<div class="footer-center">
		<a href="https://github.com/TomMakesThings">&copy; TomMakesThings <script>document.write(new Date().getUTCFullYear());</script></a>
	</div>
</div>

<script src="Javascript/graph_search.js"></script>
<script src="Javascript/navbar.js"></script>

</body>
</html>
